### ISSUE TITLE
`sisl.io.vasp.chgSileVASP.read_grid()` does not read in CHGCAR files properly

### ISSUE THREAD

**Issue by JeS24** [zerothi/sisl#472]
**Describe the bug**

I have a bunch of CHGCAR files and I want to extract the gridded data from them. To that end, I tried using the `read_grid()` method offered by `sisl` across different siles. The exact code I tried is as follows:
```python
import sisl

path = './CHGCAR'
grid = sisl.get_sile(path).read_grid()
```

This gave me the following warning & error message (some text edited for brevity):
```
<path/to/venv>/lib/python3.9/site-packages/sisl/io/vasp/chg.py:79: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  vals = np.array(vals).astype(dtype).ravel()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
TypeError: float() argument must be a string or a number, not 'list'

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<path/to/proj/dir>/data.ipynb Cell 3' in <cell line: 4>()
      1 import sisl
      3 path = './CHGCAR'
----> 4 grid = sisl.get_sile(path).read_grid()

File <path/to/venv>/lib/python3.9/site-packages/sisl/io/sile.py:514, in sile_fh_open.<locals>._wrapper.<locals>.pre_open(self, *args, **kwargs)
    512     return func(self, *args, **kwargs)
    513 with self:
--> 514     return func(self, *args, **kwargs)

File <path/to/venv>/lib/python3.9/site-packages/sisl/io/vasp/chg.py:79, in chgSileVASP.read_grid(self, index, dtype)
     76         rl()
     78 # Cut size before proceeding (otherwise it *may* fail)
---> 79 vals = np.array(vals).astype(dtype).ravel()
     80 if is_index:
     81     val = vals[n * index:n * (index+1)].reshape(nz, ny, nx)

ValueError: setting an array element with a sequence.
```
Following the error message, I found that `ChgSileVASP` is the sile-type that applies to my case, that has been defined in [this file](https://github.com/zerothi/sisl/blob/main/sisl/io/vasp/chg.py) in the source. Therein, I discovered that the lines from the CHGCAR are read sequentially and collected into a list of lists (`vals`), which is then converted to a `numpy.ndarray` in [line 79](https://github.com/zerothi/sisl/blob/64bdde7d0583d1ed5426676d86392ae6dfd16a12/sisl/io/vasp/chg.py#L79), as in the error message above. That line is the source of the aforementioned warning / error, with the cause being `vals` having lists of different lengths. In particular, the last list in `vals` can have 0 to 5 elements. For the 0-element case, everything works fine, but if the length is non-zero (a ragged sequence), `numpy.array()` cannot ascertain the dtype and so, the error is thrown. A workaround can be as simple as changing that line to the following:
```
vals = np.concatenate(vals).astype(dtype)
```
This works because `concatenate()` does not have to worry about different lengths and `ravel()` isn't needed, since the returned array is always 1D. I have tested this with all my CHGCAR files and I have not come across the same error or any other error. If the maintainers prefer, I can create a PR with this fix.

**Version details**
Run the below code and add to bug-report:
```python
import sys
print(sys.version)
import sisl
print(sisl.__version__)
```

Output:
```
3.9.9 (main, Dec 24 2021, 03:17:25) 
[GCC 11.2.0]
0.12.2
```

**Comment by zerothi** [comment]
I think it is now fixed (with a slightly different approach).

I don't have access to a non-conforming chgcar file, could you please report back if it didn't work?

**Comment by zerothi** [comment]
And thanks for the issue!

**Comment by zerothi** [comment]
Thanks for the feedback! 

> @zerothi Thanks. The fix works for all my files. And it is quite a nice fix. Removes the need to flatten or concatenate.

Indeed, the performance should be better like this ;)

Happy sisling ;)

**Comment by pfebrer** [comment]
Is the `CHGCAR` just a matrix of formatted numbers? Why not just use `np.loadtxt`? It should probably be faster, isn't it?

**Comment by zerothi** [comment]
> Is the `CHGCAR` just a matrix of formatted numbers? Why not just use `np.loadtxt`? It should probably be faster, isn't it?

It also contains other things. :(

**Comment by pfebrer** [comment]
Yes, but I mean, the grid is a section with just numbers and you know the size of it, or not? You could probably pass the file handle to `numpy` whenever it is in the grid section (?).

Anyway, if this is not the case, why not just allocate a numpy array and fill it as you go, since you seem to know the size of the grid beforehand (?). It should be more memory efficient and probably faster, no?

**Comment by zerothi** [comment]
True, I haven't played with that. I did some tests for some simple things a long time ago, and there loadtxt had some overhead, however I don't recall if I did the tests for large grids as well.

There are many places we could do better, but I think until we hit big problems time is better spent elsewhere ;)
